pennylane/devices/qubit/preprocess.py

```diff
@@ -0,0 +1,238 @@
+
+    if hamiltonian_in_obs and (not is_shadow or grouping_known):
+        # If the observable contains a Hamiltonian and if the simulation uses finite shots,
+        # or if the Hamiltonian explicitly specifies an observable grouping, split tape
+        # into multiple tapes of diagonalizable known observables.
+        try:
+            circuits, hamiltonian_fn = qml.transforms.hamiltonian_expand(circuit, group=False)
+        except ValueError as e:
+            raise ValueError(
+                "Can only return the expectation of a single Hamiltonian observable"
+            ) from e
+    elif expval_sum_in_obs and not is_shadow:
+        circuits, hamiltonian_fn = qml.transforms.sum_expand(circuit)
+
+    else:
+        # otherwise, return the output of an identity transform
+        circuits = [circuit]
+
+        def hamiltonian_fn(res):
+            return res[0]
+
+    # Check whether the circuit was broadcasted
+    if circuit.batch_size is None:
+        # If the circuit wasn't broadcasted, no action required
+        return circuits, hamiltonian_fn
+
+    # Expand each of the broadcasted Hamiltonian-expanded circuits
+    expanded_tapes, expanded_fn = qml.transforms.map_batch_transform(
+        qml.transforms.broadcast_expand, circuits
+    )
+
+    # Chain the postprocessing functions of the broadcasted-tape expansions and the Hamiltonian
+    # expansion. Note that the application order is reversed compared to the expansion order,
+    # i.e. while we first applied `hamiltonian_expand` to the tape, we need to process the
+    # results from the broadcast expansion first.
+    def total_processing(results):
+        return hamiltonian_fn(expanded_fn(results))
+
+    return expanded_tapes, total_processing
+
+
+def check_validity(tape):
+    """Checks whether the operations and observables in queue are all supported by the device.
+
+    Args:
+        tape (.QuantumTape): tape from which to validate operations and observables
+
+    Raises:
+        DeviceError: if there are operations in the queue or observables that the device does
+            not support
+    """
+    for o in tape.operations:
+        operation_name = o.name
+
+        if not _stopping_condition(o):
+            raise DeviceError(f"Gate {operation_name} not supported on Python Device")
+
+    for o in tape.observables:
+        if isinstance(o, qml.measurements.MeasurementProcess) and o.obs is not None:
+            o = o.obs
+
+        if isinstance(o, Tensor):
+            for i in o.obs:
+                if not _supports_observable(i.name):
+                    raise DeviceError(f"Observable {i.name} not supported on Python Device")
+        else:
+            observable_name = o.name
+
+            if not _supports_observable(observable_name):
+                raise DeviceError(f"Observable {observable_name} not supported on Python Device")
+
+
+def preprocess(tapes, execution_config=None, max_expansion=10):
```

Comment:
So the idea is that developers creating a plugin device can either use this `preprocess` directly, or create their own? (@albi3ro)

Reply:
This `preprocess` would be exclusively for the new `PythonDevice`, and devices that inherit from it can override `preprocess` if needed.

Reply:
what would that inheritance/override look like, given that this isn't a member function of any PythonDevice class?

Reply:
```
from .preprocess import preprocess as default_qubit_preprocessor

class DefaultQubit2(QuantumDevice):

    def preprocess(self, circuits, execution_config : ExecutionConfig = DefaultExecutionConfig):
        circuits, post_proc_fn = default_qubit_preprocessor(circuits, execution_config)
        if execution_config.gradient_method == "device":
                # expand till all trainable operations have generators
                # raise error if finite shots
                # raise error if not expectation value of observable with a matrix or sparse matrix
        if execution_config.gradient_method == "backprop":
               circuits = qml.simplify(circuits)
               # backpropogation doesn't work with `Pow`
        return circuits, post_proc_fn
```

Do we 100% need to extract it to another file and then bind it like this? Maybe not, but it will allow us to focus on documentation, integration, and binding things together in the file that defines the `DefaultQubit2` class, not smaller details.

Reply:
is "DefaultQubit2" our first new Python Device? If so, and if this is the expected use, maybe we should call this function something like `default_preprocess` since each device will have to implement their own `preprocess`, most likely using our shared (default?) code, idk. that makes a fair bit of sense at a glance though

Reply:
I think it's definitely worth renaming the function at some point, but I don't think it's a pressing issue at the moment. That being said, I think that a discussion soon would be a good idea about overridding `preprocess`. I think that if it were to be overridden for every new python/hardware device, then there is a lot of potential for repeated code.